### PR TITLE
Segment refactoring

### PR DESCRIPTION
**Pull request by Maxell92**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | 
| Automated tests included? | Y
| Related user documentation PR URL | https://github.com/mautic/documentation/pull/262
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | N
| Deprecations? | Y

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

We added new code to handle segments. Old methods are still available, but deprecated.
This PR is focused on code refactoring and to replace subSelect by JOINs in SQL. New queries should be faster then old ones in most cases.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. 
2. 

#### Steps to test this PR:
1. Make sure you have database with lot of data (segments and leads).
2. Update all your segments from console
3. Checkout this branch and clear cache.
4. Run command `app/console mautic:segments:check-builders --env=dev`. It compares old result to new ones. You can specify segment's id by `--segment-id=1` parameter if you hit limits or need to explore any segment.
5. We have fixed a few bugs in old code (e.g. handling relative dates), so results could be little different. You need to compare old SQL to new one and find if it a bug. You can find both queries in log.
6. We have several different query builders. We need least 1 test for each of these filters:
- Email read count
- Email received
- Email read date
- Hit URL count
- Hit URL date
- DNC bounced
- DNC unsubscribed
- Device type / brand
- UTM tags
7. Run command for segments update. No error should occur and No changes in segment should be found.
8. Clear all `lead_lists_leads` data and run segment update again. All data should be processed correctly.

#### List deprecations along with the new alternative:
1.  `LeadListRepository / getLeadsByList()` > `ContactSegmentService` and proper method (`getNewLeadListLeadsCount`, `getTotalLeadListLeadsCount`, `getNewLeadListLeads`, `getOrphanedLeadListLeadsCount`, `getOrphanedLeadListLeads`
2. `ListModel / getLeadsByList` > `ContactSegmentService` and proper method

#### List backwards compatibility breaks:
1. 
2.

### CODE REVIEW

**Comment by Maxell92** [comment]
Related PRs:
- https://github.com/mautic/mautic/pull/5886
- https://github.com/mautic/mautic/pull/5911

We need to solve it somehow - rewrite it for new code or close that PR.

**Comment by Maxell92** [comment]
I have included UTM filtering from https://github.com/mautic/mautic/pull/5886.

File: app/bundles/LeadBundle/Segment/Decorator/Date/Other/DateAnniversary.php [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+use Mautic\LeadBundle\Segment\Decorator\FilterDecoratorInterface;
+
+class DateAnniversary implements FilterDecoratorInterface
+{
+    /**
+     * @var DateDecorator
+     */
+    private $dateDecorator;
+
+    public function __construct(DateDecorator $dateDecorator)
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Segment/Decorator/Date/Other/DateAnniversary.php]: Missing docblock

File: app/bundles/LeadBundle/Segment/Decorator/Date/Other/DateAnniversary.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    {
+        $this->dateDecorator = $dateDecorator;
+    }
+
+    public function getField(ContactSegmentFilterCrate $contactSegmentFilterCrate)
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Segment/Decorator/Date/Other/DateAnniversary.php]: Missing docblocks for this method and those below

File: app/bundles/LeadBundle/Segment/Decorator/Date/Other/DateAnniversary.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        return $this->dateDecorator->getParameterHolder($contactSegmentFilterCrate, $argument);
+    }
+
+    public function getParameterValue(ContactSegmentFilterCrate $contactSegmentFilterCrate)
+    {
+        $dateTimeHelper = $this->dateDecorator->getDefaultDate();
+
+        return $dateTimeHelper->toUtcString('%-m-d');
+
+        return '%'.date('-m-d');
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Segment/Decorator/Date/Other/DateAnniversary.php]: Which return is correct?

File: app/bundles/LeadBundle/Segment/Decorator/Date/Other/DateDefault.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Segment/Decorator/Date/Other/DateDefault.php]: 2018

File: app/bundles/LeadBundle/Segment/Decorator/Date/Other/DateDefault.php [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+     * @param DateDecorator $dateDecorator
+     * @param string        $originalValue
+     */
+    public function __construct(DateDecorator $dateDecorator, $originalValue)
+    {
+        $this->dateDecorator = $dateDecorator;
+        $this->originalValue = $originalValue;
+    }
+
+    public function getField(ContactSegmentFilterCrate $contactSegmentFilterCrate)
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Segment/Decorator/Date/Other/DateDefault.php]: Docblocks?

File: app/bundles/LeadBundle/Segment/Decorator/Date/Other/DateRelativeInterval.php [hunk 6]
```diff
@@ -0,0 +1,95 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Segment/Decorator/Date/Other/DateRelativeInterval.php]: 2018

File: app/bundles/LeadBundle/Segment/Decorator/Date/Other/DateRelativeInterval.php [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+     * @param DateDecorator $dateDecorator
+     * @param string        $originalValue
+     */
+    public function __construct(DateDecorator $dateDecorator, $originalValue)
+    {
+        $this->dateDecorator = $dateDecorator;
+        $this->originalValue = $originalValue;
+    }
+
+    public function getField(ContactSegmentFilterCrate $contactSegmentFilterCrate)
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Segment/Decorator/Date/Other/DateRelativeInterval.php]: Docblocks?

File: app/bundles/LeadBundle/Segment/Decorator/Date/Week/DateWeekAbstract.php [hunk 8]
```diff
@@ -0,0 +1,50 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Segment/Decorator/Date/Week/DateWeekAbstract.php]: 2018

File: app/bundles/LeadBundle/Segment/Decorator/Date/Week/DateWeekLast.php [hunk 9]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Segment/Decorator/Date/Week/DateWeekLast.php]: 2018

File: app/bundles/LeadBundle/Segment/Decorator/Date/Week/DateWeekNext.php [hunk 10]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Segment/Decorator/Date/Week/DateWeekNext.php]: 2018

File: app/bundles/LeadBundle/Segment/Decorator/Date/Week/DateWeekThis.php [hunk 11]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Segment/Decorator/Date/Week/DateWeekThis.php]: 2018

File: app/bundles/LeadBundle/Segment/Decorator/Date/Year/DateYearAbstract.php [hunk 12]
```diff
@@ -0,0 +1,43 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Segment/Decorator/Date/Year/DateYearAbstract.php]: 2018

File: app/bundles/LeadBundle/Segment/Decorator/Date/Year/DateYearLast.php [hunk 13]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Segment/Decorator/Date/Year/DateYearLast.php]: 2018

File: app/bundles/LeadBundle/Segment/Decorator/Date/Year/DateYearNext.php [hunk 14]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Segment/Decorator/Date/Year/DateYearNext.php]: 2018

File: app/bundles/LeadBundle/Segment/Decorator/Date/Year/DateYearThis.php [hunk 15]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Segment/Decorator/Date/Year/DateYearThis.php]: 2018

File: app/bundles/LeadBundle/Segment/Decorator/DateDecorator.php [hunk 16]
```diff
@@ -0,0 +1,36 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Segment/Decorator/DateDecorator.php]: 2018

File: app/bundles/LeadBundle/Segment/Decorator/DateDecorator.php [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+     * @param ContactSegmentFilterCrate $contactSegmentFilterCrate
+     *
+     * @throws \Exception
+     */
+    public function getParameterValue(ContactSegmentFilterCrate $contactSegmentFilterCrate)
+    {
+        throw new \Exception('Instance of Date option needs to implement this function');
+    }
+
+    public function getDefaultDate()
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Segment/Decorator/DateDecorator.php]: docblock?

File: app/bundles/LeadBundle/Segment/Decorator/DecoratorFactory.php [hunk 18]
```diff
@@ -0,0 +1,82 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Segment/Decorator/DecoratorFactory.php]: 2018

File: app/bundles/LeadBundle/Segment/DoNotContact/DoNotContactParts.php [hunk 19]
```diff
@@ -0,0 +1,52 @@
+<?php
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Segment/DoNotContact/DoNotContactParts.php]: 2018

File: app/bundles/LeadBundle/Segment/RelativeDate.php [hunk 20]
```diff
@@ -0,0 +1,62 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Segment/RelativeDate.php]: 2018

File: app/bundles/LeadBundle/Segment/RandomParameterName.php [hunk 21]
```diff
@@ -0,0 +1,47 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Segment/RandomParameterName.php]: 2018

File: app/bundles/LeadBundle/Services/ContactSegmentFilterDictionary.php [hunk 22]
```diff
@@ -0,0 +1,222 @@
...
+use Mautic\LeadBundle\Segment\Query\Filter\DoNotContactFilterQueryBuilder;
+use Mautic\LeadBundle\Segment\Query\Filter\ForeignFuncFilterQueryBuilder;
+use Mautic\LeadBundle\Segment\Query\Filter\ForeignValueFilterQueryBuilder;
+use Mautic\LeadBundle\Segment\Query\Filter\SegmentReferenceFilterQueryBuilder;
+use Mautic\LeadBundle\Segment\Query\Filter\SessionsFilterQueryBuilder;
+
+/**
+ * Class ContactSegmentFilterDictionary.
+ *
+ * @todo @petr Já jsem to myslím předělával už. Chtěl jsem z toho pak udělat i objekt, aby se člověk nemusel ptát na klíče v poli, ale pak jsme na to nesahali, protože to nebylo komplet
```
Comment (alanhartless) [inline app/bundles/LeadBundle/Services/ContactSegmentFilterDictionary.php]: Todo here

File: parse-jmeter-output.php [hunk 23]
```diff
@@ -0,0 +1,40 @@
+<?php
```
Comment (alanhartless) [inline parse-jmeter-output.php]: What is this?

File: app/bundles/LeadBundle/Command/CheckQueryPerformanceSupplierCommand.php [hunk 24]
```diff
@@ -0,0 +1,136 @@
...
+            ->setDescription('Blah')
+            ->addOption('--segment-id', '-i', InputOption::VALUE_OPTIONAL, 'Set the ID of segment to process')
+            ->addOption('--skip-old', null, InputOption::VALUE_NONE, 'Skip old query builder');
+
+        parent::configure();
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        define('blah_2', true);
```
[the remark this answers is not in the document]
Comment (galvani) [reply app/bundles/LeadBundle/Command/CheckQueryPerformanceSupplierCommand.php]: yes, this should not be there :-(

File: app/bundles/LeadBundle/Command/CheckQueryPerformanceSupplierCommand.php [hunk 25]
```diff
@@ -0,0 +1,136 @@
...
+            }
+
+            unset($l);
+
+            unset($lists);
+        }
+
+        $total = $ok + $failed;
+        //$output->writeln('');
+        //$output->writeln(sprintf('<info>Total success rate: %d%%, %d succeeded: and <error>%s</error>%s failed...</info> ', round(($ok / $total) * 100), $ok, ($failed ? $failed : ''), (!$failed ? $failed : '')));
```
[the remark this answers is not in the document]
Comment (galvani) [reply app/bundles/LeadBundle/Command/CheckQueryPerformanceSupplierCommand.php]: Actually I have no idea where is this coming from, I don't remember writing this file and I doubt it's working.

File: app/bundles/LeadBundle/Entity/LeadListRepository.php [hunk 26]
```diff
@@ -508,7 +516,28 @@ public function getLeadsByList($lists, $args = [])
...
+                        $val = "'$val'";
+                    }
+                    if (is_array($val)) {
+                        $val = join(',', $val);
+                    }
+                    $sqlT                    = str_replace(":{$key}", $val, $sqlT);
+                }
+
+                $logger->debug(sprintf('Old version SQL: %s', $sqlT));
+                if (defined('blah_2')) {
```
[the remark this answers is not in the document]
Comment (galvani) [reply app/bundles/LeadBundle/Entity/LeadListRepository.php]: this seems to be related to the benchmarking command. We have still not addressed the speed of the queries. But I have mentioned this like 100 times already so I won't no more.

File: app/bundles/LeadBundle/Model/ListModel.php [hunk 27]
```diff
@@ -996,6 +1060,11 @@ public function rebuildListLeads(LeadList $entity, $limit = 1000, $maxLeads = fa
         return $leadsProcessed;
     }
 
+    public function rebuildListLeads()
+    {
+        throw new  \Exception('Deprecated function, use updateLeadList instead');
```
[the remark this answers is not in the document]
Comment (galvani) [reply app/bundles/LeadBundle/Model/ListModel.php]: Yes we should

File: app/bundles/LeadBundle/Segment/ContactSegmentService.php [hunk 28]
```diff
@@ -0,0 +1,347 @@
...
+    }
+
+    /**
+     * @param LeadList $segment
+     *
+     * @return array
+     *
+     * @throws \Exception
+     *
+     * @nottotodo This is almost copy of getNewLeadListLeadsCount method. Only difference is that it calls getTotalSegmentContactsQuery
```
[the remark this answers is not in the document]
Comment (galvani) [reply app/bundles/LeadBundle/Segment/ContactSegmentService.php]: Oops, yes it is. I will remove it.

File: app/bundles/LeadBundle/Segment/ContactSegmentService.php [hunk 29]
```diff
@@ -0,0 +1,347 @@
...
+    }
+
+    /**
+     * Formatting helper.
+     *
+     * @param $inputSeconds
+     *
+     * @return string
+     */
+    private function format_period($inputSeconds)
```
[the remark this answers is not in the document]
Comment (galvani) [reply app/bundles/LeadBundle/Segment/ContactSegmentService.php]: There is a reason but no good one. It's copy&paste job and lazy programmer.

File: app/bundles/LeadBundle/Segment/Decorator/Date/DateOptionAbstract.php [hunk 30]
```diff
@@ -0,0 +1,135 @@
...
+    /**
+     * This function returns an operator if between range is needed. Could return like or between.
+     *
+     * @param ContactSegmentFilterCrate $leadSegmentFilterCrate
+     *
+     * @return string
+     */
+    abstract protected function getOperatorForBetweenRange(ContactSegmentFilterCrate $leadSegmentFilterCrate);
+
+    public function getField(ContactSegmentFilterCrate $contactSegmentFilterCrate)
```
[the remark this answers is not in the document]
Comment (galvani) [reply app/bundles/LeadBundle/Segment/Decorator/Date/DateOptionAbstract.php]: @Maxell92 could you, please?

File: app/bundles/LeadBundle/Model/ListModel.php [hunk 31]
```diff
@@ -763,50 +811,90 @@ public function getGlobalLists()
...
-     * @return int
+     * @throws \Exception
      */
-    public function rebuildListLeads(LeadList $entity, $limit = 1000, $maxLeads = false, OutputInterface $output = null)
+    public function getVersionNew(LeadList $entity)
     {
-        defined('MAUTIC_REBUILDING_LEAD_LISTS') or define('MAUTIC_REBUILDING_LEAD_LISTS', 1);
-
         $id       = $entity->getId();
         $list     = ['id' => $id, 'filters' => $entity->getFilters()];
```
[the remark this answers is not in the document]
Comment (galvani) [reply app/bundles/LeadBundle/Model/ListModel.php]: well, I guess the getVersionNew is meant for the debugging. These functions should no make to the release, but am I to remove them while we still don't have optimized queries?

File: app/bundles/LeadBundle/Segment/Decorator/Date/DateOptionAbstract.php [hunk 32]
```diff
@@ -0,0 +1,135 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
[the remark this answers is not in the document]
Comment (galvani) [reply app/bundles/LeadBundle/Segment/Decorator/Date/DateOptionAbstract.php]: @Maxell92 your time travel

File: app/bundles/LeadBundle/Command/CheckQueryPerformanceSupplierCommand.php (hunk 24 above)
[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/LeadBundle/Command/CheckQueryPerformanceSupplierCommand.php]: Removed.

File: app/bundles/LeadBundle/Command/CheckQueryPerformanceSupplierCommand.php (hunk 25 above)
[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/LeadBundle/Command/CheckQueryPerformanceSupplierCommand.php]: I have removed whole class.

File: app/bundles/LeadBundle/Entity/LeadListRepository.php (hunk 26 above)
[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/LeadBundle/Entity/LeadListRepository.php]: Removed

File: app/bundles/LeadBundle/Segment/Decorator/FilterDecoratorInterface.php [hunk 33]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+namespace Mautic\LeadBundle\Segment\Decorator;
+
+use Mautic\LeadBundle\Segment\ContactSegmentFilterCrate;
+
+/**
+ * Interface FilterDecoratorInterface.
+ *
+ * @TODO @petr document these functions please with phpdoc and meaningful description
```
[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/LeadBundle/Segment/Decorator/FilterDecoratorInterface.php]: Done

File: app/bundles/LeadBundle/Segment/ContactSegmentService.php [hunk 34]
```diff
@@ -0,0 +1,347 @@
...
+     * @var Logger
+     */
+    private $logger;
+
+    /**
+     * @var QueryBuilder
+     */
+    private $preparedQB;
+
+    public function __construct(
```
[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/LeadBundle/Segment/ContactSegmentService.php]: Addressed

File: app/bundles/LeadBundle/Segment/ContactSegmentService.php (hunk 29 above)
[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/LeadBundle/Segment/ContactSegmentService.php]: Removed

File: app/bundles/LeadBundle/Segment/Decorator/Date/DateOptionAbstract.php (hunk 32 above)
[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/LeadBundle/Segment/Decorator/Date/DateOptionAbstract.php]: Fixed

File: app/bundles/LeadBundle/Segment/Decorator/Date/DateOptionAbstract.php (hunk 30 above)
[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/LeadBundle/Segment/Decorator/Date/DateOptionAbstract.php]: Added

File: app/bundles/LeadBundle/Segment/Decorator/Date/DateOptionFactory.php [hunk 35]
```diff
@@ -0,0 +1,104 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/LeadBundle/Segment/Decorator/Date/DateOptionFactory.php]: Fixed

File: app/bundles/LeadBundle/Segment/Decorator/Date/DateOptionParameters.php [hunk 36]
```diff
@@ -0,0 +1,103 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/LeadBundle/Segment/Decorator/Date/DateOptionParameters.php]: Fixed

File: app/bundles/LeadBundle/Segment/Decorator/Date/Day/DateDayAbstract.php [hunk 37]
```diff
@@ -0,0 +1,43 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/LeadBundle/Segment/Decorator/Date/Day/DateDayAbstract.php]: Fixed

File: app/bundles/LeadBundle/Segment/Decorator/Date/Day/DateDayToday.php [hunk 38]
```diff
@@ -0,0 +1,24 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/LeadBundle/Segment/Decorator/Date/Day/DateDayToday.php]: Fixed

File: app/bundles/LeadBundle/Segment/Decorator/Date/Day/DateDayTomorrow.php [hunk 39]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/LeadBundle/Segment/Decorator/Date/Day/DateDayTomorrow.php]: Fixed

File: app/bundles/LeadBundle/Segment/Decorator/Date/Month/DateMonthAbstract.php [hunk 40]
```diff
@@ -0,0 +1,43 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/LeadBundle/Segment/Decorator/Date/Month/DateMonthAbstract.php]: Fixed

File: app/bundles/LeadBundle/Segment/Decorator/Date/Month/DateMonthLast.php [hunk 41]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/LeadBundle/Segment/Decorator/Date/Month/DateMonthLast.php]: Fixed

File: app/bundles/LeadBundle/Segment/Decorator/Date/Month/DateMonthNext.php [hunk 42]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/LeadBundle/Segment/Decorator/Date/Month/DateMonthNext.php]: Fixed

File: app/bundles/LeadBundle/Segment/Decorator/Date/Month/DateMonthThis.php [hunk 43]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/LeadBundle/Segment/Decorator/Date/Month/DateMonthThis.php]: Fixed

File: app/bundles/LeadBundle/Segment/Decorator/Date/Other/DateAnniversary.php [hunk 44]
```diff
@@ -0,0 +1,73 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/LeadBundle/Segment/Decorator/Date/Other/DateAnniversary.php]: Fixed

File: app/bundles/LeadBundle/Segment/Decorator/Date/Other/DateAnniversary.php (hunk 1 above)
Comment (Maxell92) [reply app/bundles/LeadBundle/Segment/Decorator/Date/Other/DateAnniversary.php]: Fixed

File: app/bundles/LeadBundle/Segment/Decorator/Date/Other/DateAnniversary.php (hunk 2 above)
Comment (Maxell92) [reply app/bundles/LeadBundle/Segment/Decorator/Date/Other/DateAnniversary.php]: Fixed

File: app/bundles/LeadBundle/Entity/LeadListRepository.php [hunk 45]
```diff
@@ -311,24 +312,32 @@ public function getLeadCount($listIds)
...
-     * @param       $lists
-     * @param array $args
+     * @param        $lists
+     * @param array  $args
+     * @param Logger $logger
      *
      * @return array
      */
-    public function getLeadsByList($lists, $args = [])
+    public function getLeadsByList($lists, $args = [], Logger $logger)
```
[the remark this answers is not in the document]
Comment (alanhartless) [reply app/bundles/LeadBundle/Entity/LeadListRepository.php]: If necessary, default it to null then create a NullLogger if === null

File: app/bundles/LeadBundle/Command/CheckQueryPerformanceSupplierCommand.php (hunk 24 above)
[the remark this answers is not in the document]
Comment (galvani) [reply app/bundles/LeadBundle/Command/CheckQueryPerformanceSupplierCommand.php]: What is removed? Either remove whole the command or nothingg

File: app/bundles/LeadBundle/Command/CheckQueryPerformanceSupplierCommand.php (hunk 25 above)
[the remark this answers is not in the document]
Comment (galvani) [reply app/bundles/LeadBundle/Command/CheckQueryPerformanceSupplierCommand.php]: Why is it still here?

[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/LeadBundle/Command/CheckQueryPerformanceSupplierCommand.php]: The class was removed. This comment is mark as outdated.

File: app/bundles/LeadBundle/Command/CheckQueryPerformanceSupplierCommand.php (hunk 24 above)
[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/LeadBundle/Command/CheckQueryPerformanceSupplierCommand.php]: The class was removed. This comment is mark as outdated.

File: app/bundles/LeadBundle/Entity/LeadListRepository.php (hunk 45 above)
[the remark this answers is not in the document]
Comment (Maxell92) [reply app/bundles/LeadBundle/Entity/LeadListRepository.php]: Resolved

**Comment by kuzmany** [comment]
@Maxell92 Can I use service for this feature request - Support relative dates for campaign action update contact  https://github.com/mautic/mautic/issues/6101 ?
I think it's attached to segments and I have to write my own services to translate relative dates to datetime.

**Comment by Maxell92** [comment]
Hello @kuzmany,

I think you need another class. Relative dates are supposed to translate strings like `+1 day`, `-6 months`, `this week` to DateTime instance/s. It has nothing to do with date formatting.

**Comment by javjim-mautic** [comment]
Awesome work, testing done, working properly

**Pull request merged**